Infinispan Protostream Marshaller

pruivo opened the pull request:
Closes #29394

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

I'm gathering earlier feedback and sharing how it will end up.
Pending tasks are:

- [x] Compute the final Protostream IDs. With Protostream IDs, it generates smaller payloads since it sends the IDs (Goole Protobuf varInt) instead of the message name.
- [x] Refactor events and predicates. I'm taking the opportunity to remove unused fields and duplicated code.
- [x] Migration from JBoss Marshalling to Infinispan Protostream
- [x] Re-add JBoss Marshalling to allow migration of sessions
- [x] Add clearing of caches on migration to KC25 - all entries in the caches will be gone
- [x] Documentation for the release notes

Out of scope: Existing entries in the session cache will be migrated in PR https://github.com/keycloak/keycloak/pull/29377

### Migration

The migration path is under discussion and has the following assumptions:

* Only user and client sessions will be migrated. All other data stored in distributed/replicated caches will be lost.
* Embedded cache needs to have persistence configured. If only in-memory storage is configured, we assume that losing sessions during Keycloak major version upgrade is acceptable and expected.

The most appealing solution is to start a separate, non-clustered Infinispan `CacheManager` with JBoss-Marshalling configured (the `CacheManager` is limited to a single marshalled). This `CacheManager` will be used to read the old session data. Then, we have the following 2 scenarios

* Persistent User Sessions feature is enabled. Using the JBoss-Marshalling `CacheManager` to read the sessions and store them in the database. The cache is clear at the end.
* Without Persistent User Session feature, it reads the sessions from the JBoss-Marshalling `CacheManager` and writes them in the Protostream enabled `CacheManager`.

pruivo committed:
Infinispan Protostream Marshaller

Closes #29394

Signed-off-by: Pedro Ruivo <pruivo@redhat.com>
Co-authored-by: Alexander Schwartz <aschwart@redhat.com>

ryanemerson commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/authorization/events/ResourceEvent.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    public String toString() {
+        return String.format("ResourceRemovedEvent [ id=%s, name=%s, eventType=%s]", id, name, eventType());
```
`s/ResourceRemovedEvent/ResourceEvent`

ryanemerson commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/events/TypedInvalidationEvent.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Proto
+    public enum Type {
+        UPDATED, REMOVED, ADDED
```
From a schema evolution POV I think it would make more sense to keep the three event types as distinct messages. That way the schema is more representative of the three logic types and we can easily add new event specific fields to the messages if required.

Obviously there's a lot of duplicated code between these classes, but we could still leverage an abstract class to reduce this.

pruivo replied:
I was afraid someone would say that :rofl: 
Sure, I can revert my changes.

ryanemerson reviewed (commented):
@pruivo Do you plan to add the proto compatibility check capabilities as part of this PR or should we create an issue to add that in a subsequent PR? I would be happy with the latter given how large the current PR is.

stianst commented:
Please do not merge until we decide if this should be included in Keycloak 25, or shortly after the release

ryanemerson commented on docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:
```diff
@@ -0,0 +1,20 @@
+bin/kc.sh build --features=persistent-user-session ...
+----
+
+This process will import all the session from in-memory Infinispan caches into the database.
```
Should we be using the `{jdgserver_name}` placeholder here, and in your other docs?

pruivo replied:
I replaced in this line but left `Infinispan Protostream` unchanged since "protostream" is not a product so the name stays the same.

ahus1 commented:
@pruivo - I've pushed one change for the docs: 

* Those docs shouldn't link to GitHub issues
* I don't want to repeat contents from the KC25 migration guide, instead I'd rather point to it to avoid inconsistencies
*

ahus1 commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java:
```diff
@@ -1015,8 +1015,12 @@ public static UUID createClientSessionUUID(String userSessionId, String clientId
-            migrateNonPersistentSessionsToPersistentSessions();
+        // Changed encoding from JBoss Marshalling to ProtoStream.
+        // Unable to read the cached data.
+        if ("25.0.0".equals(modelVersion) || "26.0.0".equals(modelVersion)) {
+            InfinispanConnectionProvider.DISTRIBUTED_REPLICATED_CACHE_NAMES.stream()
+                    .map(cacheName -> session.getProvider(InfinispanConnectionProvider.class).getCache(cacheName))
+                    .forEach(Cache::clear);
```
There seems to be a misunderstanding of how this works, and this is most likely to the API I created here. 

This `migrate()` is not called on every version, but only when it is called explicitly. The only place it is called is in `MigrateTo25_0_0.java`, and there it is always called with `25.0.0`. 

To have the behavior of always clearing the caches for the upgrade to KC26, the following would be necessary:
* Create a class `MigrationTo26_0_0.java and register it so it is called
* Then call the `migrate()` method.

In some earlier version in this or another PR we had a setup where we would open the Infinispan configuration with the JBoss Marshaller and then migrate the sessions. If we're not doing this, we wouldn't allow people to upgrade from KC24 to KC26 in one go, which would mean that a smooth migration wouldn't be available for RHBK users as there is no RHBK25. 

If we choose to not support the straight migration, the code in `PersistentUserSessionProvider.migrate()` should change as follows: 
* Update the code in `PersistentUserSessionProvider` to have a NOOP for `25.0.0`
* Clear the caches for `26.0.0` as the current code shows

Still, this clearing of the caches seems to be a bit unrelated to `PersistentUserSessionProvider` as it is clearing several other caches outside of the user sessions. Maybe the InfinispanConnectionProvider() should be doing this kind of work?

pruivo replied:
In the meeting last month, we agreed on not performing a migration since if users want to save their sessions, they should enable persistent sessions in KC25. Are we changing our minds? :)

ahus1 replied:
I have to admit I wasn't sure and possibly got confused with the snipped that was starting the second cache manager. If we then still agree on that we don't offer a migration for RHBK, then so be it. 

Then the only bit remains where to call the clearing of caches.

ahus1 reviewed (changes requested):
Hi @pruivo - thank you for the PR. See below for a question around migrations and how we want to handle them. 

In additional to that, a merge conflict appeared :-( ... please rebase and resolve. Thanks!

pruivo commented:
Summary of what we agreed:
* Create `MigrateTo26_0_0.java` class
* Clear the clustered caches
* Move the migration code to `InfinispanConnectionProvider` or a more generic provider
* Remove `migrate()` from `PersistentUserSessionProvider` since it won't work.

About the RHBK users, we may need to provide an offline tool to migrate the content of the external Infinispan into the correct encoding. Adding the logic to handle 2 different encodings would be complex and would have to live in the Keycloak server code for a long time.

mhajas commented on server-spi/src/main/java/org/keycloak/component/ComponentModel.java:
This still implements `Serializable`, is it correct?

pruivo replied:
I think it can be removed. I'm unsure if it is even stored in any cache at all.

pruivo replied:
I pushed a commit to try and I've removed the class from protostream. Let's see if the test suite works or not.

mhajas replied:
Let's see. I would expect it fails since it is used here: https://github.com/pruivo/keycloak/blob/t_protostream_ispn15/model/storage-private/src/main/java/org/keycloak/storage/managers/UserStorageSyncManager.java#L386

pruivo replied:
you're right :+1:

mhajas reviewed (commented):
Thank you for the changes @pruivo. I haven't found any blockers or unexpected changes. I added one comment to the code.

pruivo commented:
Ready to review :partying_face: 

I could not use `InfinispanConnectionProvider` for the migration since that API is not visible from `MigrateTo26_0_0.java`. It can be changed in a future PR or if you have a simpler alternative for this PR, I could do the change.

cc @ahus1 @mhajas @ryanemerson

ahus1 reviewed (dismissed):
Thank you for the PR! Found a typo in the include for the release notes. Used the opportunity to remove some now unused code in the PersitentUserSessionProvider.

pruivo commented:
> Thank you for the PR! Found a typo in the include for the release notes. Used the opportunity to remove some now unused code in the PersitentUserSessionProvider.

I didn't remove that method because of the javadoc: `This method is public so people can use it to build their custom migrations or re-import sessions when necessary in a future version of Keycloak.`

Isn't useful anymore?

ahus1 commented:
> Isn't useful anymore?

With the change of the serialization format, it IMHO became less useful. But ok, let's keep it for now. I realize I only deleted half of it anyway :-(

ryanemerson commented on model/infinispan/src/main/java/org/keycloak/marshalling/Marshalling.java:
```diff
@@ -1,25 +1,155 @@
 
-    // Note: Min ID is 2500
-    public static final Integer REPLACE_FUNCTION_ID = 2500;
+    // Model
```
We can add a test to ensure that all of the ID values in this class are unique, in the Infinispan repo we do:

```java

   @Test
   public void testIdUniqueness() throws Exception {
      Class clazz = ProtoStreamTypeIds.class;
      Field[] fields = clazz.getFields();
      Set<Integer> messageIds = new HashSet<>();
      Set<Integer> lowerBounds = new HashSet<>();
      for (Field f : fields) {
         if (f.getName().endsWith("_LOWER_BOUND"))
            assertTrue(f.getName(), lowerBounds.add(f.getInt(clazz)));
         else
            assertTrue(f.getName(), messageIds.add(f.getInt(clazz)));
      }
      assertTrue(!messageIds.isEmpty());
      assertTrue(!lowerBounds.isEmpty());
      assertEquals(fields.length - lowerBounds.size(), messageIds.size());
      assertEquals(fields.length - messageIds.size(), lowerBounds.size());
   }
```

In this case the test will be simpler as we don't define LOWER_BOUND fields.

pruivo replied:
It is not required. If you have a duplicate ID, the build will fail with 

`org.infinispan.protostream.DescriptorParserException: Duplicate type id 65541 for type keycloak.LockEntry. Already used by keycloak.LockEntryPredicate`

ryanemerson replied:
I'm obviously behind on the times. We can probably remove that test from Infinispan now then :slightly_smiling_face:

ryanemerson reviewed (commented):
Great work @pruivo, just a few minor things I have noticed so far.

pruivo commented:
Added a new commit with @ryanemerson suggestions :+1:

pruivo commented on model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/events/InvalidationEvent.java:
```diff
@@ -17,13 +17,15 @@
  */
 public abstract class InvalidationEvent implements ClusterEvent {
 
+    @ProtoField(1)
```
done

pruivo commented on model/infinispan/src/main/java/org/keycloak/marshalling/Marshalling.java:
```diff
@@ -1,25 +1,155 @@
+    // see org.keycloak.storage.UserStorageProviderModel
+    public static final int USER_STORAGE_PROVIDER_MODES = 65539;
+    // see org.keycloak.storage.managers.UserStorageSyncManager.UserStorageProviderClusterEvent
+    public static final int USER_STORAGE_PROVIDER_CLUSTER_EVENT = 65540;
```
The modules where they are used do not depend on this one. The comments tell where they are used.

pruivo commented:
Rebased and squashed the commits. 

I suspect that `CacheInitializer#loadSessions()` is running concurrently with `UserSessionProvider#migrate()`. The migrate method will clear the cache and remove the `InitializerState` preventing nodes to start. I couldn't get any logs from GitHub Actions to confirm this.

I changed how the migration works and moved the code to `DefaultInfinispanConnectionProvider` to avoid code duplication. It no longers clears the in-memory data (assumes there isn't a mix of KC25 and KC26 servers in the same cluster) but only from persistence (which is the only place where the data is stored in binary). 

cc @ahus1 @mhajas @ryanemerson

mhajas reviewed (approved):
Thank you very much @pruivo! This was quite a journey. Let's see what the rest of it will look like! It looks good to me.
